### PR TITLE
Add TreeCache recipe

### PR DESCRIPTION
**Pull request by tevino**
A port of [Apache Curator's TreeCache recipe](http://curator.apache.org/curator-recipes/tree-cache.html)

### CODE REVIEW

**Comment by tevino** [comment]
@basvanbeek Please review.

File: api.go [hunk 1]
```diff
@@ -1,58 +1,58 @@
-package curator
```
Comment (wooparadog) [inline api.go]: 文件一模一样也有 diff 呃... 是 `set ff=unix` 了么?

File: recipes/cache/tree_cache.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+// publishEvent publish an event with given type and data to all listeners.
+func (tc *TreeCache) publishEvent(tp TreeCacheEventType, data *ChildData) {
+	if tc.state.Value() != curator.STOPPED {
+		evt := TreeCacheEvent{Type: tp, Data: data}
+		tc.logger.Debugf("publishEvent: %v", evt)
+		tc.callListeners(evt)
+	}
+}
```
Comment (tonyseek) [inline recipes/cache/tree_cache.go:300]: Maybe we should call the user listeners in a background goroutine? It is the default behavior of Apache Curator (via `executorService.submit(new Runnable() {})`) and it prevents listeners from blocking each other.

File: create.go [hunk 3]
```diff
@@ -1,141 +1,141 @@
...
+			return "", err
+		} else {
+			payload = data
+		}
+	}
+
+	adjustedPath := b.client.fixForNamespace(givenPath, b.createMode.IsSequential())
+
+	if b.backgrounding.inBackground {
+		go b.pathInBackground(adjustedPath, payload, givenPath)
```
Comment (tonyseek) [inline create.go:32]: The background operations are consumed from a blocking queue in the implementation of Apache Curator. Maybe we should make the background goroutines be sequential too?

File: api.go (hunk 1 above)
Comment (tevino) [inline api.go]: gofmt

File: recipes/cache/tree_cache.go (hunk 2 above)
Comment (tevino) [inline recipes/cache/tree_cache.go:300]: Right.

File: create.go (hunk 3 above)
Comment (tevino) [inline create.go:32]: The change was made by `gofmt`, I didn't intent to modify this part, not in this PR.

But I guess this should be done.

Comment (tonyseek) [inline create.go:32]: gotcha

**Comment by basvanbeek** [comment]
@tevino: thank you for your PR. I used goimports to reformat the code as it was long overdue. Can you rebase your code so I get a clean diff?

**Comment by tevino** [comment]
I have some in-progress commits not pushed yet, I'll rebase later, meanwhile you can add `?w=1` after the URL for a clean diff.

**Comment by tevino** [comment]
@basvanbeek Rebased

**Comment by tevino** [comment]
@basvanbeek Rebased

**Comment by basvanbeek** [comment]
@tevino just a ping to let you know it's still on the radar. Need to finish a deadline. Will look at request early next week. Is it complete or do you still have commits you want to push?

**Comment by tevino** [comment]
@basvanbeek glad to hear that, actually I'm also busy for some other things recently, I do have commits about goroutine safety to finish, won't be too large I guess.

**Comment by tevino** [comment]
@basvanbeek Races of this recipe are fixed, please review.

There are still races in the implementation of curator itself, I fixed some of them here, however further changes should be in another PR.

I'll squash the commits if that's preferred.

**Comment by basvanbeek** [comment]
thanks @tevino for the recipe.

**Comment by brian-brazil** [comment]
There's a battle-tested TreeCache at https://github.com/prometheus/prometheus/blob/master/util/treecache/treecache.go which we believe to be race free if you want to start from there.

**Comment by tevino** [comment]
The recipe @brian-brazil provided above is lack of important features like **path selector** and **event listener**, the behaviour of error handling of a `conn.GetW` is hard-coded to be sending deleting events(special handling of prometheus?), while this PR contains a nearly full port of the original TreeCache recipe together with its test cases.

I'm sorry but honestly I don't believe https://github.com/prometheus/prometheus/blob/master/util/treecache/treecache.go is a good start for general or robust application.

**Pull request merged**